Improved charset handling

skimberk opened the pull request:
I was generating a fuzzer for the JSON grammar from `antlr/grammars-v4` and noticed some differences between how ANTLR and grammarinator handle charsets.

The specific cases I noticed:

`["\\\u0000-\u001F]` represents the characters `"` and `\` and the range `\u0000-\u001F`. However, `re.split(r'(\w-\w)', src)` doesn't split as intended so grammarinator parses this as five characters.

`[+\-]` represents two characters (`+` and `-`) in ANTLR, with `-` being escaped. However, this escape isn't handled by `.decode('unicode_escape')`, so grammarinator parses it as `+`, `\`, and `-`.

I ported a simplified version of ANTLR's charset parsing code to Python to replace the current handling. It handles the above cases correctly, as well as adds the ability to specify unicode using the `\u{123}` syntax, which would fix https://github.com/renatahodovan/grammarinator/issues/15.

The new code is considerably more complex than the previous code, so I put it in a new file. I'm not sure if this is within the scope of changes you think are useful, or if it should be implemented some other way, but I figured I'd make the PR and see what you think.

akosthekiss commented on tests/grammars/Charset.g4:
```diff
@@ -0,0 +1,39 @@
+
+NON_RANGE_DASH
+	: [-a] [a-] [-]
+	;
```
Some also interesting char sets:
- `[--]` (ANTLR warns about '-' being used multiple times, but eventually accepts the char set. '-' twice.)
- `[-\-]` (Again, ANTLR warning, but still OK. Same as above, '-' twice, but one dash escaped, the other not.)
- `[-\--]` (Warning, but still OK. '-' 3x. The middle dash is _not_ a range marker.)
- `[---]` (No warning! This is a totally fine range of '-'..'-', i.e., a dash.)
- `[\--\-]` (Same as above, but this is how it should look like according to ANTLR docs.)

(The "tricky" part in [ANTLR's documentation](https://github.com/antlr/antlr4/blob/master/doc/lexer-rules.md#lexer-rule-elements)  is that it states that "To get ], \, or - you must escape them with \.", but that's actually not true. If the dash is the first or the last character in the char set, it does not need to be escaped. Which leads to the above interestingly valid patterns.)

KvanTTT replied:
Yes, documentation is outdated. A hyphen `-` is treated as a range operator being placed between two chars. Otherwise, it's treated as an ordinary hyphen. If you want to declare a hyphen between two chars, you should use escaping.

akosthekiss commented:
Hi @skimberk, are you willing to follow up on this, addressing the reviews?

skimberk commented:
Sorry, I've been busy the last couple weeks. I'll make updates in response
to your feedback, but it might be a couple weeks before I have the chance
to (I don't currently have access to my computer).

Thanks so much for taking the time to consider my PR and give feedback!

On Sat, Jun 13, 2020, 02:25 Akos Kiss <notifications@github.com> wrote:

> Hi @skimberk <https://github.com/skimberk>, are you willing to follow up
> on this, addressing the reviews?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/renatahodovan/grammarinator/pull/20#issuecomment-643584056>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAKS66745K7UNLRXSFT7ZU3RWMSVNANCNFSM4MQIOTRQ>
> .
>

skimberk commented on grammarinator/charset_parser.py:
```diff
@@ -0,0 +1,119 @@
+
+            if in_range:
+                in_range = False
+                ranges[-1] = (ranges[-1][0], codepoint + 1)
```
Tried this, got `TypeError: 'tuple' object does not support item assignment`. I guess it's because tuples are immutable.

akosthekiss replied:
Indeed, I've missed the fact that `ranges` contains tuples. My bad, ignore this comment.

skimberk commented on grammarinator/charset_parser.py:
```diff
@@ -0,0 +1,119 @@
+    if escaped == 'u':
+        if offset + 3 > len(s):
+            # \u{1} is the shortest we support
+            raise EscapeParseError('Shortest unicode escape is \\u{0}')
```
It ended up being `if hex_start_offset == hex_end_offset:`

skimberk commented:
Rebased onto latest master and squashed all the previous commits (left one commit to show most recent changes). I fixed everything you mentioned in the comments, with the only two modifications being the two comments I responded to.

Sorry for the very long delay, I didn't have access to my computer. Further responses will be faster now that I'm back home.

akosthekiss commented on grammarinator/process.py:
```diff
@@ -230,25 +231,6 @@ def character_range_interval(node):
         return (int(start.replace('\\u', '0x'), 16) if '\\u' in start else ord(start),
                 int(end.replace('\\u', '0x'), 16) if '\\u' in end else ord(end) + 1)
 
-    def lexer_charset_interval(src):
-        elements = re.split(r'(\w-\w)', src)
-        ranges = []
-        for element in elements:
-            if not element:
-                continue
-
-            # Convert character sequences like \n, \t, etc. into a single character.
-            element = bytes(element, 'utf-8').decode('unicode_escape')
-            if len(element) > 1:
-                if element[1] == '-' and len(element) == 3:
-                    ranges.append((ord(element[0]), ord(element[2]) + 1))
-                else:
-                    for char in element:
-                        ranges.append((ord(char), ord(char) + 1))
-            elif len(element) == 1:
-                ranges.append((ord(element), ord(element) + 1))
-        return ranges
```
How about having the new implementation here, replacing the original function. Something like the code below. (Disclaimer: the suggestion may be a bit rough around the edges. It has not been tested, not even executed. "Should work (TM)".)

```py
def lexer_charset_char(s, offset):
    # To be kept in sync with org.antlr.v4.misc.EscapeSequenceParsing.parseEscape
    if s[offset] != '\\':
        return ord(s[offset]), offset + 1

    if offset + 2 > len(s):
        raise ValueError('Escape must have at least two characters')

    escaped = s[offset + 1]
    offset += 2  # Move past backslash and escaped character

    if escaped == 'u':
        if s[offset] == '{':
            # \u{...}
            hex_start_offset = offset + 1
            hex_end_offset = s.find('}', hex_start_offset)
            if hex_end_offset == -1:
                raise ValueError('Missing closing bracket for unicode escape')
            if hex_start_offset == hex_end_offset:
                raise ValueError('Missing codepoint for unicode escape')

            offset = hex_end_offset + 1 # Skip over last bracket
        else:
            # \uXXXX
            hex_start_offset = offset
            hex_end_offset = hex_start_offset + 4
            if hex_end_offset > len(s):
                raise ValueError('Non-bracketed unicode escape must be of form \\uXXXX')

            offset = hex_end_offset

        try:
            codepoint = int(s[hex_start_offset:hex_end_offset], 16)
        except ValueError:
            raise ValueError('Invalid hex value')

        if codepoint < 0 or codepoint > sys.maxunicode:
            raise ValueError('Invalid unicode codepoint')

        return codepoint, offset

    if escaped in ('p', 'P'):
        raise ValueError('Unicode properties (\\p{...}) are not supported')

    if escaped in escaped_values:
        # To be kept in sync with org.antlr.v4.misc.CharSupport.ANTLRLiteralEscapedCharValue
        return ord({
            'n': '\n',
            'r': '\r',
            'b': '\b',
            't': '\t',
            'f': '\f',
            '\\': '\\',
            # Additional escape sequences defined by org.antlr.v4.misc.EscapeSequenceParsing.parseEscape
            '-': '-',
            ']': ']',
        }[escaped]), offset

    raise ValueError('Invalid escaped value')


def lexer_charset_interval(s):
    # To be kept in sync with org.antlr.v4.automata.LexerATNFactory.getSetFromCharSetLiteral
    assert len(s) > 0, 'Charset cannot be empty'

    ranges = []

    offset = 0
    while offset < len(s):
        in_range = s[offset] == '-' and offset != 0 and offset != len(s) - 1
        if in_range:
            offset += 1

        codepoint, offset = lexer_charset_char(s, offset)

        if in_range:
            ranges[-1] = (ranges[-1][0], codepoint + 1)
        else:
            ranges.append((codepoint, codepoint + 1))

    return ranges
```

skimberk commented:
Updated with your suggestions pretty much verbatim. The only changes:

- kept the `escaped_values` object, although I moved it into the function
- kept comment about Python's handling of multi-character unicode codepoints
- added extra spaces in front of inline comments for linter

I made the changes in two extra commits. I made the changes in the first, then moved the functions into `process.py` in the second. Let me know what you think and I can squash. Thanks!

skimberk commented:
Just squashed. Looks like I don't have permissions to merge, so I'll leave that to you. Thanks again for all your feedback!
